pallets/providers/src/utils.rs

```diff
@@ -999,6 +999,165 @@ where
             .saturating_mul(accrued_failed_submission_count)
             .saturating_mul(2u32.into()))
     }
+
+    /// Adjust the fixed rate payment stream between a user and an MSP based on the [`RateDeltaParam`].
+    ///
+    /// Handles creating, updating, or deleting the fixed rate payment stream storage.
+    fn apply_delta_fixed_rate_payment_stream(
+        msp_id: &MainStorageProviderId<T>,
+        bucket_id: &BucketId<T>,
+        user_id: &T::AccountId,
+        delta: RateDeltaParam<T>,
+    ) -> Result<(), DispatchError> {
+        let current_rate = <T::PaymentStreams as PaymentStreamsInterface>::get_inner_fixed_rate_payment_stream_value(
+            &msp_id,
+            &user_id,
+        )
+        .unwrap_or_default();
+
+        match delta {
+            RateDeltaParam::NewBucket => {
+                let bucket: Bucket<T> =
+                    Buckets::<T>::get(&bucket_id).ok_or(Error::<T>::BucketNotFound)?;
+
+                let new_rate = if bucket.size.is_zero() {
+                    current_rate
+                        .checked_add(&T::ZeroSizeBucketFixedRate::get())
+                        .ok_or(Error::<T>::Overflow)?
+                } else {
+                    let value_prop = MainStorageProviderIdsToValuePropositions::<T>::get(
+                        &msp_id,
+                        &bucket.value_prop_id,
+                    )
+                    .ok_or(Error::<T>::ValuePropositionNotFound)?;
+
+                    let bucket_rate =
+                        value_prop.price_per_unit_of_data_per_block * bucket.size.into();
+
+                    current_rate
+                        .checked_add(&bucket_rate)
+                        .ok_or(Error::<T>::Overflow)?
+                };
```

Comment:
I would have thought that the `new_rate` would be calculated as
```rust
let new_rate = current_rate + &T::ZeroSizeBucketFixedRate::get() + value_prop * bucket_size;
```

Otherwise it might be cheaper to have a bucket with some small size, than have it with no size at all.